### PR TITLE
Remove unused parser declarations and unused parser type declarations that result from `RemoveRedundantParsers` pass.

### PR DESCRIPTION
**Pull request by kfcripps**
This fixes the following `Compiler Bug` that occurs when building the attached test:
```
In file: /home/kyle/repos/p4c/frontends/p4/inlining.cpp:522
Compiler Bug: /home/kyle/repos/p4c/frontends/p4/inlining.cpp:522: Null firstCall
```

Please see the test comments for details.

### CODE REVIEW

File: frontends/p4/redundantParsers.cpp [hunk 1]
```diff
@@ -37,6 +37,11 @@ bool FindRedundantParsers::preorder(const IR::P4Parser *parser) {
 }
 
 const IR::Node *EliminateSubparserCalls::postorder(IR::MethodCallStatement *mcs) {
+    // Only remove redundant parser applications in parser states. Parser applications
+    // made in illegal locations (such as control blocks) will be detected by and error
+    // will be reported in later passes (i.e. DiscoverInlining).
```
Comment (vlstill) [inline frontends/p4/redundantParsers.cpp:42]: It would be great if you also add test that covers this case.

Comment (kfcripps) [reply frontends/p4/redundantParsers.cpp:42]: We already have test [issue1331.p4](https://github.com/p4lang/p4c/blob/main/testdata/p4_16_errors_outputs/issue1331.p4-stderr) that will fail without this change.

**Review by vlstill — APPROVED** [review]
Makes sense to me.

**Comment by kfcripps** [comment]
Rebased. Please merge when CI passes.

**Comment by kfcripps** [comment]
Rebased again

File: frontends/p4/frontend.cpp [hunk 2]
```diff
@@ -228,8 +228,8 @@ const IR::P4Program *FrontEnd::run(const CompilerOptions &options, const IR::P4P
         new RemoveReturns(&refMap),
         new RemoveDontcareArgs(&refMap, &typeMap),
         new MoveConstructors(&refMap),
-        new RemoveAllUnusedDeclarations(&refMap),
         new RemoveRedundantParsers(&refMap, &typeMap),
+        new RemoveAllUnusedDeclarations(&refMap),
```
[the remark this answers is not in the document]
Comment (kfcripps) [reply frontends/p4/frontend.cpp]: Done.

**Comment by kfcripps** [comment]
```
99% tests passed, 2 tests failed out of 2729

Label Time Summary:
XFAIL                              =  36.85 sec*proc (34 tests)
bmv2                               = 1708.43 sec*proc (566 tests)
bmv2-parser-inline-opt-disabled    =  37.83 sec*proc (13 tests)
bmv2-parser-inline-opt-enabled     =  39.69 sec*proc (13 tests)
bmv2-ptf                           =  14.27 sec*proc (1 test)
dpdk                               = 313.53 sec*proc (249 tests)
ebpf                               =  31.30 sec*proc (28 tests)
ebpf-bcc                           =   8.20 sec*proc (26 tests)
ebpf-errors                        =   0.25 sec*proc (1 test)
ebpf-kernel                        = 192.72 sec*proc (30 tests)
err                                =  42.19 sec*proc (315 tests)
graph                              =   0.81 sec*proc (2 tests)
gtest                              =  41.77 sec*proc (1 test)
p14_to_16                          = 419.25 sec*proc (194 tests)
p4                                 = 1341.15 sec*proc (1216 tests)
p4tc                               =  24.35 sec*proc (38 tests)
p4unroll                           =  17.50 sec*proc (13 tests)
ubpf                               =  13.22 sec*proc (19 tests)

Total Test time (real) = 1063.01 sec

The following tests FAILED:
Errors while running CTest
	545 - bmv2-parser-inline-opt-disabled/testdata/p4_16_samples/parser-inline/parser-inline-test1.p4 (Failed)
	571 - bmv2-ptf/testdata/p4_16_samples/ternary2-bmv2.p4 (Failed)
Error: Process completed with exit code 8.
```
These tests did not fail locally. Can someone please re-run the failing job?

**Comment by kfcripps** [comment]
Rebased.

**Comment by kfcripps** [comment]
It looks like a test is spuriously failing, so please ignore the failure or re-trigger the job. Thanks.

**Comment by kfcripps** [comment]
@vlstill @mihaibudiu Also I can't merge, so feel free to rebase (if needed) and merge this if you have no other comments.

**Review by vlstill — APPROVED** [review]

**Comment by kfcripps** [comment]
@vlstill @mihaibudiu Thanks!

**Pull request merged**